### PR TITLE
Refine icd10who & icd10cm regexes and distinctions

### PR DESCRIPTION
**Pull request by dhimmel**
Currently, `icd10:C34.90` fails but `icd10:C34.9` succeeds.

### CODE REVIEW

**Comment by dhimmel** [comment]
https://github.com/biopragmatics/bioregistry/pull/559/commits/a3264bfd32a28811be90ff0df22d03093f4b6e72 triggers the error on [CI](https://github.com/biopragmatics/bioregistry/runs/8279036707?check_suite_focus=true#step:5:120)... excellent:

```
_______________________ TestRegistry.test_extra_examples _______________________
E   AssertionError: False is not true : Failed on prefix=icd10: C34.90
```

**Comment by dhimmel** [comment]
Ah wait, there's way more that choke the regex:

- C91.Z
- C91.Z0
- C4A.8
- D47.Z2

**Comment by cthoyt** [comment]
Interesting you're looking into this at the moment, I had just queued https://github.com/bioregistry/regex-report to rerun, which will give some more ICD10 examples when it finishes (right now it's spending a lot of time on UMLS...)

**Comment by dhimmel** [comment]
Nice. Will await your report on failing ICD10 identifiers before tweaking the regex.

**Comment by cthoyt** [comment]
So here are 296 of ICD10 that weren't passing the regex (this wasn't taking into account your improvement)

```yaml
- invalid: 296
  invalid_percent: 12.62
  invalid_sample:
  - identifier: F90-F98
    link: https://icd.who.int/browse10/2019/en#/F90-F98
    name: Behavioural and emotional disorders with onset usually occurring in childhood
      and adolescence
  - identifier: T08-T14
    link: https://icd.who.int/browse10/2019/en#/T08-T14
    name: Injuries to unspecified part of trunk, limb or body region
  - identifier: X50-X57
    link: https://icd.who.int/browse10/2019/en#/X50-X57
    name: Overexertion, travel and privation
  - identifier: G90-G99
    link: https://icd.who.int/browse10/2019/en#/G90-G99
    name: Other disorders of the nervous system
  - identifier: XX
    link: https://icd.who.int/browse10/2019/en#/XX
    name: External causes of morbidity and mortality
  - identifier: B85-B89
    link: https://icd.who.int/browse10/2019/en#/B85-B89
    name: Pediculosis, acariasis and other infestations
  - identifier: R20-R23
    link: https://icd.who.int/browse10/2019/en#/R20-R23
    name: Symptoms and signs involving the skin and subcutaneous tissue
  - identifier: I95-I99
    link: https://icd.who.int/browse10/2019/en#/I95-I99
    name: Other and unspecified disorders of the circulatory system
  - identifier: L55-L59
    link: https://icd.who.int/browse10/2019/en#/L55-L59
    name: Radiation-related disorders of the skin and subcutaneous tissue
  - identifier: K40-K46
    link: https://icd.who.int/browse10/2019/en#/K40-K46
    name: Hernia
  - identifier: Z55-Z65
    link: https://icd.who.int/browse10/2019/en#/Z55-Z65
    name: Persons with potential health hazards related to socioeconomic and psychosocial
      circumstances
  - identifier: Z40-Z54
    link: https://icd.who.int/browse10/2019/en#/Z40-Z54
    name: Persons encountering health services for specific procedures and health
      care
  - identifier: Y83-Y84
    link: https://icd.who.int/browse10/2019/en#/Y83-Y84
    name: Surgical and other medical procedures as the cause of abnormal reaction
      of the patient, or of later complication, without mention of misadventure at
      the time of the procedure
  - identifier: B99-B99
    link: https://icd.who.int/browse10/2019/en#/B99-B99
    name: Other infectious diseases
  - identifier: V95-V97
    link: https://icd.who.int/browse10/2019/en#/V95-V97
    name: Air and space transport accidents
  - identifier: Y90-Y98
    link: https://icd.who.int/browse10/2019/en#/Y90-Y98
    name: Supplementary factors related to causes of morbidity and mortality classified
      elsewhere
  - identifier: P05-P08
    link: https://icd.who.int/browse10/2019/en#/P05-P08
    name: Disorders related to length of gestation and fetal growth
  - identifier: J30-J39
    link: https://icd.who.int/browse10/2019/en#/J30-J39
    name: Other diseases of upper respiratory tract
  - identifier: V40-V49
    link: https://icd.who.int/browse10/2019/en#/V40-V49
    name: Car occupant injured in transport accident
  - identifier: M20-M25
    link: https://icd.who.int/browse10/2019/en#/M20-M25
    name: Other joint disorders
  - identifier: J00-J06
    link: https://icd.who.int/browse10/2019/en#/J00-J06
    name: Acute upper respiratory infections
  - identifier: A65-A69
    link: https://icd.who.int/browse10/2019/en#/A65-A69
    name: Other spirochaetal diseases
  - identifier: XVII
    link: https://icd.who.int/browse10/2019/en#/XVII
    name: Congenital malformations, deformations and chromosomal abnormalities
  - identifier: X00-X09
    link: https://icd.who.int/browse10/2019/en#/X00-X09
    name: Exposure to smoke, fire and flames
  - identifier: XIX
    link: https://icd.who.int/browse10/2019/en#/XIX
    name: Injury, poisoning and certain other consequences of external causes
  - identifier: R40-R46
    link: https://icd.who.int/browse10/2019/en#/R40-R46
    name: Symptoms and signs involving cognition, perception, emotional state and
      behaviour
  - identifier: X85-Y09
    link: https://icd.who.int/browse10/2019/en#/X85-Y09
    name: Assault
  - identifier: M15-M19
    link: https://icd.who.int/browse10/2019/en#/M15-M19
    name: Arthrosis
  - identifier: U82-U85
    link: https://icd.who.int/browse10/2019/en#/U82-U85
    name: Resistance to antimicrobial and antineoplastic drugs
  - identifier: C60-C63
    link: https://icd.who.int/browse10/2019/en#/C60-C63
    name: Malignant neoplasms of male genital organs
  - identifier: O60-O75
    link: https://icd.who.int/browse10/2019/en#/O60-O75
    name: Complications of labour and delivery
  - identifier: C30-C39
    link: https://icd.who.int/browse10/2019/en#/C30-C39
    name: Malignant neoplasms of respiratory and intrathoracic organs
  - identifier: M60-M63
    link: https://icd.who.int/browse10/2019/en#/M60-M63
    name: Disorders of muscles
  - identifier: X
    link: https://icd.who.int/browse10/2019/en#/X
    name: Diseases of the respiratory system
  - identifier: R25-R29
    link: https://icd.who.int/browse10/2019/en#/R25-R29
    name: Symptoms and signs involving the nervous and musculoskeletal systems
  - identifier: J40-J47
    link: https://icd.who.int/browse10/2019/en#/J40-J47
    name: Chronic lower respiratory diseases
  - identifier: Y85-Y89
    link: https://icd.who.int/browse10/2019/en#/Y85-Y89
    name: Sequelae of external causes of morbidity and mortality
  - identifier: H40-H42
    link: https://icd.who.int/browse10/2019/en#/H40-H42
    name: Glaucoma
  - identifier: IV
    link: https://icd.who.int/browse10/2019/en#/IV
    name: Endocrine, nutritional and metabolic diseases
  - identifier: V90-V94
    link: https://icd.who.int/browse10/2019/en#/V90-V94
    name: Water transport accidents
  - identifier: M70-M79
    link: https://icd.who.int/browse10/2019/en#/M70-M79
    name: Other soft tissue disorders
  - identifier: XVIII
    link: https://icd.who.int/browse10/2019/en#/XVIII
    name: Symptoms, signs and abnormal clinical and laboratory findings, not elsewhere
      classified
  - identifier: A15-A19
    link: https://icd.who.int/browse10/2019/en#/A15-A19
    name: Tuberculosis
  - identifier: Q60-Q64
    link: https://icd.who.int/browse10/2019/en#/Q60-Q64
    name: Congenital malformations of the urinary system
  - identifier: Q20-Q28
    link: https://icd.who.int/browse10/2019/en#/Q20-Q28
    name: Congenital malformations of the circulatory system
  - identifier: M91-M94
    link: https://icd.who.int/browse10/2019/en#/M91-M94
    name: Chondropathies
  - identifier: O00-O08
    link: https://icd.who.int/browse10/2019/en#/O00-O08
    name: Pregnancy with abortive outcome
  - identifier: V30-V39
    link: https://icd.who.int/browse10/2019/en#/V30-V39
    name: Occupant of three-wheeled motor vehicle injured in transport accident
  - identifier: H80-H83
    link: https://icd.who.int/browse10/2019/en#/H80-H83
    name: Diseases of inner ear
  - identifier: L20-L30
    link: https://icd.who.int/browse10/2019/en#/L20-L30
    name: Dermatitis and eczema
  - identifier: F20-F29
    link: https://icd.who.int/browse10/2019/en#/F20-F29
    name: Schizophrenia, schizotypal and delusional disorders
  - identifier: J20-J22
    link: https://icd.who.int/browse10/2019/en#/J20-J22
    name: Other acute lower respiratory infections
  - identifier: E40-E46
    link: https://icd.who.int/browse10/2019/en#/E40-E46
    name: Malnutrition
  - identifier: F60-F69
    link: https://icd.who.int/browse10/2019/en#/F60-F69
    name: Disorders of adult personality and behaviour
  - identifier: III
    link: https://icd.who.int/browse10/2019/en#/III
    name: Diseases of the blood and blood-forming organs and certain disorders involving
      the immune mechanism
  - identifier: U00-U49
    link: https://icd.who.int/browse10/2019/en#/U00-U49
    name: Provisional assignment of new diseases of uncertain etiology or emergency
      use
  - identifier: F30-F39
    link: https://icd.who.int/browse10/2019/en#/F30-F39
    name: Mood [affective] disorders
  - identifier: VIII
    link: https://icd.who.int/browse10/2019/en#/VIII
    name: Diseases of the ear and mastoid process
  - identifier: P50-P61
    link: https://icd.who.int/browse10/2019/en#/P50-P61
    name: Haemorrhagic and haematological disorders of fetus and newborn
  - identifier: K90-K93
    link: https://icd.who.int/browse10/2019/en#/K90-K93
    name: Other diseases of the digestive system
  - identifier: G50-G59
    link: https://icd.who.int/browse10/2019/en#/G50-G59
    name: Nerve, nerve root and plexus disorders
  - identifier: G70-G73
    link: https://icd.who.int/browse10/2019/en#/G70-G73
    name: Diseases of myoneural junction and muscle
  - identifier: F00-F09
    link: https://icd.who.int/browse10/2019/en#/F00-F09
    name: Organic, including symptomatic, mental disorders
  - identifier: Q10-Q18
    link: https://icd.who.int/browse10/2019/en#/Q10-Q18
    name: Congenital malformations of eye, ear, face and neck
  - identifier: C69-C72
    link: https://icd.who.int/browse10/2019/en#/C69-C72
    name: Malignant neoplasms of eye, brain and other parts of central nervous system
  - identifier: Q35-Q37
    link: https://icd.who.int/browse10/2019/en#/Q35-Q37
    name: Cleft lip and cleft palate
  - identifier: J90-J94
    link: https://icd.who.int/browse10/2019/en#/J90-J94
    name: Other diseases of pleura
  - identifier: G35-G37
    link: https://icd.who.int/browse10/2019/en#/G35-G37
    name: Demyelinating diseases of the central nervous system
  - identifier: VI
    link: https://icd.who.int/browse10/2019/en#/VI
    name: Diseases of the nervous system
  - identifier: T26-T28
    link: https://icd.who.int/browse10/2019/en#/T26-T28
    name: Burns and corrosions confined to eye and internal organs
  - identifier: F80-F89
    link: https://icd.who.int/browse10/2019/en#/F80-F89
    name: Disorders of psychological development
  - identifier: M65-M68
    link: https://icd.who.int/browse10/2019/en#/M65-M68
    name: Disorders of synovium and tendon
  - identifier: E65-E68
    link: https://icd.who.int/browse10/2019/en#/E65-E68
    name: Obesity and other hyperalimentation
  - identifier: H60-H62
    link: https://icd.who.int/browse10/2019/en#/H60-H62
    name: Diseases of external ear
  - identifier: C50-C50
    link: https://icd.who.int/browse10/2019/en#/C50-C50
    name: Malignant neoplasm of breast
  name: International Classification of Diseases, 10th Revision
  pattern: ^[A-Z]\d+(\.[-\d+])?$
  prefix: icd10
  total: 2345
  version: null
```

**Comment by dhimmel** [comment]
Turns out `C34.90` is not actually in ICD-10, but only ICD-10-CM! So it is probably actually correct that `ICD10:C34.90` is invalid.

Looking through https://icd.who.int/browse10/2019/en, there are a few different types of `ICD10` (WHO version i.e. not CM) identifiers:

- top-level terms which are a roman numeral like `XVIII`.
- ranges of codes like `C00-C97` or its child `C30-C39`
- diagnosis codes, which are covered by the current regex as well as the one [described here](https://www.johndcook.com/blog/2019/05/05/regex_icd_codes/): `[A-TV-Z][0-9][0-9AB]\.?[0-9A-TV-Z]{0,4}`

So proposal is to add more examples to the proper prefix, icd10 / icd10cm. Then adjust / add regexes.

**Comment by dhimmel** [comment]
Essential readings at

- https://github.com/biopragmatics/bioregistry/issues/251
- https://github.com/biopragmatics/bioregistry/issues/485

**Comment by cthoyt** [comment]
Thanks @dhimmel for updating the regex. Maybe @joeflack4 and @matentzn can weigh in before we merge this

File: src/bioregistry/data/bioregistry.json [hunk 1]
```diff
@@ -32814,8 +32820,14 @@
...
     },
-    "description": "ICD-10-CM is the official system of assigning codes to diagnoses and procedures associated with hospital utilization in the United States.",
+    "description": "ICD-10-CM is the official system of assigning codes to diagnoses and procedures associated with hospital utilization in the United States issued by the National Center for Health Statistics (NCHS) and Centers for Medicare & Medicaid Services (CMS).",
     "example": "F4521",
+    "example_extras": [
+      "C34.90",
+      "C4A.8",
+      "C91.Z",
+      "C91.Z0"
+    ],
```
Comment (cthoyt) [inline src/bioregistry/data/bioregistry.json:32830]: you can also add a field `"example_decoys"` that has negative examples

File: src/bioregistry/data/bioregistry.json [hunk 2]
```diff
@@ -32837,6 +32849,7 @@
       "wikidata": "P4229"
     },
     "name": "International Classification of Diseases, 10th Revision, Clinical Modification",
+    "pattern": "^[A-TV-Z][0-9][0-9AB]\\.?[0-9A-TV-Z]{0,4}$",
```
Comment (dhimmel) [inline src/bioregistry/data/bioregistry.json:32852]: Might need to refine this. See https://medium.com/@manabu.torii/regex-pattern-for-icd-10-cm-codes-5763bd66e26d

File: src/bioregistry/data/bioregistry.json [hunk 3]
```diff
@@ -32837,6 +32867,7 @@
       "wikidata": "P4229"
     },
     "name": "International Classification of Diseases, 10th Revision, Clinical Modification",
+    "pattern": "^([A-Z][0-9][0-9AB]((-[A-Z][0-9][0-9AB])|(\\.[0-9A-KXZ]([0-9A-EXYZ]([0-9A-HX][0-59A-HJKMNP-S]?)?)?)?))$",
```
Comment (dhimmel) [inline src/bioregistry/data/bioregistry.json:32870]: The ICD10CM pattern is based on the one at https://medium.com/@manabu.torii/regex-pattern-for-icd-10-cm-codes-5763bd66e26d by Manabu Torii. It's based on an analysis of all 2021 ICD-10-CM and therefore might be too restrictive for future releases that use additional characters for new codes.

But can possibly deal with breakage then to get the benefits of a more restrictive regex now.

Added support for code ranges like `C00-D49`

**Comment by dhimmel** [comment]
> Maybe @joeflack4 and @matentzn can weigh in before we merge this

That would be excellent. Should be reviewable as of acc35657b5d12a18cb0626eeea9c703fe5316f4d

File: src/bioregistry/data/bioregistry.json (hunk 1 above)
Comment (matentzn) [reply src/bioregistry/data/bioregistry.json:32830]: super cool idea

File: src/bioregistry/data/bioregistry.json (hunk 3 above)
Comment (matentzn) [reply src/bioregistry/data/bioregistry.json:32870]: If it it works for the 2021 and all preceding versions I have no objections.

**Comment by matentzn** [comment]
I am generally fine. I am not going to rerun the analysis to ensure all prefixes truly match - if they don't, I will come back to relax the regex :P

**Comment by cthoyt** [comment]
I just re-ran it, and everything is passing. Will merge now 🚀

**Comment by joeflack4** [comment]
I was out yesterday. Sorry I couldn't review.

**Pull request merged**